api/include/opentelemetry/baggage/baggage.h

```diff
@@ -18,276 +19,61 @@ namespace baggage
-      {
-        first = false;
-      }
-      header_s.append(UrlEncode(key));
-      header_s.push_back(kKeyValueSeparator);
-
-      // extracting metadata from value. We do not encode metadata
-      auto metadata_separator = value.find(kMetadataSeparator);
-      if (metadata_separator != std::string::npos)
-      {
-        header_s.append(UrlEncode(value.substr(0, metadata_separator)));
-        auto metadata = value.substr(metadata_separator);
-        header_s.append(std::string(metadata.data(), metadata.size()));
-      }
-      else
-      {
-        header_s.append(UrlEncode(value));
-      }
-      return true;
-    });
-    return header_s;
-  }
-
-private:
-  static bool IsPrintableString(nostd::string_view str)
-  {
-    for (const auto ch : str)
-    {
-      if (ch < ' ' || ch > '~')
-      {
-        return false;
-      }
-    }
-
-    return true;
-  }
-
-  static bool IsValidKey(nostd::string_view key) { return key.size() && IsPrintableString(key); }
-
-  static bool IsValidValue(nostd::string_view value) { return IsPrintableString(value); }
-
-  // Uri encode key value pairs before injecting into header
-  // Implementation inspired from : https://golang.org/src/net/url/url.go?s=7851:7884#L264
-  static std::string UrlEncode(nostd::string_view str)
-  {
-    auto to_hex = [](char c) -> char {
-      static const char *hex = "0123456789ABCDEF";
-      return hex[c & 15];
-    };
-
-    std::string ret;
-
-    for (auto c : str)
-    {
-      if (std::isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~')
-      {
-        ret.push_back(c);
-      }
-      else if (c == ' ')
-      {
-        ret.push_back('+');
-      }
-      else
-      {
-        ret.push_back('%');
-        ret.push_back(to_hex(c >> 4));
-        ret.push_back(to_hex(c & 15));
-      }
-    }
-
-    return ret;
-  }
-
-  // Uri decode key value pairs after extracting from header
-  static std::string UrlDecode(nostd::string_view str, bool &err)
-  {
-    auto IsHex = [](char c) {
-      return std::isdigit(c) || (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f');
-    };
-
-    auto from_hex = [](char c) -> char {
-      return std::isdigit(c) ? c - '0' : std::toupper(c) - 'A' + 10;
-    };
-
-    std::string ret;
-
-    for (size_t i = 0; i < str.size(); i++)
-    {
-      if (str[i] == '%')
-      {
-        if (i + 2 >= str.size() || !IsHex(str[i + 1]) || !IsHex(str[i + 2]))
-        {
-          err = 1;
-          return "";
-        }
-        ret.push_back(from_hex(str[i + 1]) << 4 | from_hex(str[i + 2]));
-        i += 2;
-      }
-      else if (str[i] == '+')
-      {
-        ret.push_back(' ');
-      }
-      else if (std::isalnum(str[i]) || str[i] == '-' || str[i] == '_' || str[i] == '.' ||
-               str[i] == '~')
-      {
-        ret.push_back(str[i]);
-      }
-      else
-      {
-        err = 1;
-        return "";
-      }
-    }
-
-    return ret;
-  }
+  std::string ToHeader() const { return baggage_impl_->ToHeader(); }
```

Comment:
Since this is in API, is this `std::string` type ever going to be exposed across DLL / Shared Library boundary? Even if it's a private method, isn't this still something where we'd rather return a `nostd::string_view` instead of `std::string` ?

Reply:
It's a non-virtual function, so it should be safe to return `std::string`, it won't break ABI stability. Though we can change it as below had it been a virtual function:

```
  void ToHeader(nostd::string &header) {
}
```